### PR TITLE
Add code action provider for quick fix (repair scenario)

### PR DESCRIPTION
**Pull request by mcbelton**
Add code action provider support for quick fix repair formula scenario.

### CODE REVIEW

File: src/libraries/Microsoft.PowerFx.LanguageServerProtocol/LanguageServer/Constants.cs [hunk 1]
```diff
@@ -0,0 +1,13 @@
+﻿// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+namespace Microsoft.PowerFx.LanguageServerProtocol
+{
+    internal class Constants
+    {
+        public class CodeActionKind
```
Comment (lesaltzm) [inline src/libraries/Microsoft.PowerFx.LanguageServerProtocol/LanguageServer/Constants.cs:8]: no need for this inner classing. If you want this to be referenced as `LanguageServerProtocol.Constants.CodeActionKind.QuickFix` make a Constants folder and namespace

File: src/libraries/Microsoft.PowerFx.LanguageServerProtocol/Protocol/CodeAction.cs [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+        public CodeAction()
+        {
+            Diagnostics = new Diagnostic[] { };
+        }
+
+        public string Title { get; set; }
+
+        public string Kind { get; set; }
+
+        /// <summary>
```
Comment (lesaltzm) [inline src/libraries/Microsoft.PowerFx.LanguageServerProtocol/Protocol/CodeAction.cs:17]: ideally add descriptive comments for all these members

Comment (lesaltzm) [reply src/libraries/Microsoft.PowerFx.LanguageServerProtocol/Protocol/CodeAction.cs:17]: or links to the LSP docs

File: src/libraries/Microsoft.PowerFx.LanguageServerProtocol/Protocol/CodeActionContext.cs [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+
+namespace Microsoft.PowerFx.LanguageServerProtocol.Protocol
+{
+    /// <summary>
+    /// 
+    /// </summary>
+    public class CodeActionContext
+    {
+        /// <summary>
+        /// ctor
```
Comment (lesaltzm) [inline src/libraries/Microsoft.PowerFx.LanguageServerProtocol/Protocol/CodeActionContext.cs:12]: Can this comment be more useful?

Comment (mcbelton) [reply src/libraries/Microsoft.PowerFx.LanguageServerProtocol/Protocol/CodeActionContext.cs:12]: Yes, addressed.

File: src/libraries/Microsoft.PowerFx.LanguageServerProtocol/Protocol/CodeAction.cs (hunk 2 above)
Comment (mcbelton) [reply src/libraries/Microsoft.PowerFx.LanguageServerProtocol/Protocol/CodeAction.cs:29]: yes simplified.

File: src/libraries/Microsoft.PowerFx.LanguageServerProtocol/LanguageServer/Constants.cs (hunk 1 above)
Comment (mcbelton) [reply src/libraries/Microsoft.PowerFx.LanguageServerProtocol/LanguageServer/Constants.cs]: yes simplified.

File: src/libraries/Microsoft.PowerFx.LanguageServerProtocol/Protocol/CodeActionCommand.cs [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+
+        /// <summary>
+        /// Gets or sets command to be executed.
+        /// </summary>
+        public string Command { get; set; }
+
+        /// <summary>
+        /// Gets or sets command arguments.
+        /// </summary>
+        public object[] Arguments { get; set; }
```
Comment (MikeStall) [inline src/libraries/Microsoft.PowerFx.LanguageServerProtocol/Protocol/CodeActionCommand.cs:32]: >object [](http://example.com/codeflow?start=15&length=6)

can this be a stronger type than object?

File: src/libraries/Microsoft.PowerFx.LanguageServerProtocol/Protocol/WorkspaceEdit.cs [hunk 5]
```diff
@@ -0,0 +1,26 @@
...
+        /// </summary>
+        public WorkspaceEdit()
+        {
+            Changes = new Dictionary<string, TextEdit[]>();
+        }
+
+        /// <summary>
+        /// Gets or sets suggested changes.
+        /// </summary>
+        public Dictionary<string, TextEdit[]> Changes { get; set; }
```
Comment (MikeStall) [inline src/libraries/Microsoft.PowerFx.LanguageServerProtocol/Protocol/WorkspaceEdit.cs:24]: >string [](http://example.com/codeflow?start=26&length=6)

what is the key?

File: src/libraries/Microsoft.PowerFx.Core/Public/IPowerFxScope.cs [hunk 6]
```diff
@@ -21,5 +21,12 @@ public interface IPowerFxScope
         /// Provide intellisense for expression
         /// </summary>
         IIntellisenseResult Suggest(string expression, int cursorPosition);
+
+        /// <summary>
+        /// Provider quick fix suggesions.
+        /// </summary>
+        /// <param name="expression">The formula expression.</param>
+        /// <returns>Collection of quick fixes.</returns>
+        QuickFixResult[] QuickFix(string expression);
```
Comment (MikeStall) [inline src/libraries/Microsoft.PowerFx.Core/Public/IPowerFxScope.cs:30]: >expression [](http://example.com/codeflow?start=41&length=10)

does this need a cursorPosition too? (like intellisense?)

Comment (MikeStall) [inline src/libraries/Microsoft.PowerFx.Core/Public/IPowerFxScope.cs:30]: >QuickFixResult [](http://example.com/codeflow?start=8&length=14)

This looks like it populates the menu of available fixes, but once they select one of the quick fixes, how does it convey the actual text update to apply?

Comment (mcbelton) [reply src/libraries/Microsoft.PowerFx.Core/Public/IPowerFxScope.cs:30]: PROSE repair engine doesn't need cursor position as it needs whole expression for repair suggestion and going to return entire formula to replace with. As per my understanding might not be useful at this moment.

File: src/libraries/Microsoft.PowerFx.LanguageServerProtocol/Protocol/WorkspaceEdit.cs (hunk 5 above)
Comment (mcbelton) [reply src/libraries/Microsoft.PowerFx.LanguageServerProtocol/Protocol/WorkspaceEdit.cs:24]: key is document uri https://microsoft.github.io/language-server-protocol/specifications/specification-3-17/#workspaceEdit

File: src/libraries/Microsoft.PowerFx.Core/Public/IPowerFxScope.cs (hunk 6 above)
Comment (mcbelton) [reply src/libraries/Microsoft.PowerFx.Core/Public/IPowerFxScope.cs:30]: This is executed by via code action command (monaco editor).

File: src/libraries/Microsoft.PowerFx.LanguageServerProtocol/Protocol/CodeActionCommand.cs (hunk 4 above)
Comment (mcbelton) [reply src/libraries/Microsoft.PowerFx.LanguageServerProtocol/Protocol/CodeActionCommand.cs:32]: Removed as no need to expose.

Comment (MikeStall) [reply src/libraries/Microsoft.PowerFx.LanguageServerProtocol/Protocol/CodeActionCommand.cs:32]: I see the comment, but no code changes. Are the latest changes pushed?

Comment (MikeStall) [reply src/libraries/Microsoft.PowerFx.LanguageServerProtocol/Protocol/CodeActionCommand.cs:32]: Referred to CodeAction class.

Comment (MikeStall) [reply src/libraries/Microsoft.PowerFx.LanguageServerProtocol/Protocol/CodeActionCommand.cs:32]: See [https://docs.microsoft.com/en-us/dotnet/api/microsoft.visualstudio.languageserver.protocol.command.arguments?view=visualstudiosdk-2019#Microsoft_VisualStudio_LanguageServer_Protocol_Command_Arguments](https://docs.microsoft.com/en-us/dotnet/api/microsoft.visualstudio.languageserver.protocol.command.arguments?view=visualstudiosdk-2019#Microsoft_VisualStudio_LanguageServer_Protocol_Command_Arguments)

File: src/libraries/Microsoft.PowerFx.Core/Public/QuickFixResult.cs [hunk 7]
```diff
@@ -0,0 +1,21 @@
...
+    {
+        /// <summary>
+        /// Gets or sets title to be displayed on quick fix suggestion.
+        /// </summary>
+        public string Title { get; set; }
+
+        /// <summary>
+        /// Gets or sets quick fix expression text to be applied.
+        /// </summary>
+        public string Text { get; set; }
```
Comment (MikeStall) [inline src/libraries/Microsoft.PowerFx.Core/Public/QuickFixResult.cs:19]: >Text [](http://example.com/codeflow?start=22&length=4)

This is a full replacement of the entire expression, so no range is provided.

File: src/libraries/Microsoft.PowerFx.LanguageServerProtocol/LanguageServer/LanguageServer.cs [hunk 8]
```diff
@@ -279,9 +282,72 @@ private void HandleInitialFixupRequest(string id, string paramsJson)
...
+
+                        foreach (var item in result)
+                        {
+                            items.Add(new CodeAction()
+                            {
+                                Title = item.Title,
+                                Kind = codeActionKind,
+                                Edit = new WorkspaceEdit
+                                {
+                                    Changes = new Dictionary<string, TextEdit[]> { { documentUri, new[] { new TextEdit { Range = codeActionParams.Range, NewText = item.Text } } } }
```
Comment (MikeStall) [inline src/libraries/Microsoft.PowerFx.LanguageServerProtocol/LanguageServer/LanguageServer.cs:331]: > Range = codeActionParams.Range, [](http://example.com/codeflow?start=120&length=32)

note: this is where we need to update to have more specific ranges on the code fix.

Comment (mcbelton) [reply src/libraries/Microsoft.PowerFx.LanguageServerProtocol/LanguageServer/LanguageServer.cs]: Now, I have updated object model to return the location of fix.

File: src/libraries/Microsoft.PowerFx.LanguageServerProtocol/Protocol/WorkspaceEdit.cs (hunk 5 above)
Comment (lesaltzm) [reply src/libraries/Microsoft.PowerFx.LanguageServerProtocol/Protocol/WorkspaceEdit.cs:24]: add that as a comment in the code?

File: src/libraries/Microsoft.PowerFx.Core/Public/CodeActionRange.cs [hunk 9]
```diff
@@ -0,0 +1,31 @@
+﻿// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+namespace Microsoft.PowerFx.Core.Public
+{
+    /// <summary>
+    /// Code action fix range location.
+    /// </summary>
+    public class CodeActionRange
```
Comment (MikeStall) [inline src/libraries/Microsoft.PowerFx.Core/Public/CodeActionRange.cs:9]: >CodeActionRange [](http://example.com/codeflow?start=17&length=15)

Can we use our existing Span object?

File: src/libraries/Microsoft.PowerFx.Core/Public/IPowerFxScope.cs [hunk 10]
```diff
@@ -21,5 +21,12 @@ public interface IPowerFxScope
         /// Provide intellisense for expression
         /// </summary>
         IIntellisenseResult Suggest(string expression, int cursorPosition);
+
+        /// <summary>
+        /// Provider quick fix suggesions.
+        /// </summary>
+        /// <param name="expression">The formula expression.</param>
+        /// <returns>Collection of quick fixes.</returns>
+        CodeActionResult[] QuickFix(string expression);
```
Comment (MikeStall) [inline src/libraries/Microsoft.PowerFx.Core/Public/IPowerFxScope.cs:30]: Adding a new method to an interface is a breaking change. 
Can you put on a new interface (like what we did for IPowerFxScopeDisplayName)?

That also gives a clear way for things that don't provide this feature - they just don't implement the interface.

File: src/libraries/Microsoft.PowerFx.LanguageServerProtocol/LanguageServer/LanguageServer.cs [hunk 11]
```diff
@@ -279,9 +282,85 @@ private void HandleInitialFixupRequest(string id, string paramsJson)
             }));
         }
 
+        private void HandleCodeActionRequest(string id, string paramsJson)
```
Comment (MikeStall) [inline src/libraries/Microsoft.PowerFx.LanguageServerProtocol/LanguageServer/LanguageServer.cs:285]: >HandleCodeActionRequest [](http://example.com/codeflow?start=21&length=23)

What user gesture causes this to get called? 
(ie, how frequently does it get called...)

Comment (mcbelton) [reply src/libraries/Microsoft.PowerFx.LanguageServerProtocol/LanguageServer/LanguageServer.cs:285]: It gets called only when error occurs (conditionally not always)

File: src/libraries/Microsoft.PowerFx.Core/Public/CodeActionRange.cs (hunk 9 above)
Comment (mcbelton) [reply src/libraries/Microsoft.PowerFx.Core/Public/CodeActionRange.cs]: Refactored to move new quickfix interface under protocol project.

**Review by MikeStall — APPROVED** [review]
:shipit:

**Review by lesaltzm — APPROVED** [review]
🚢

**Pull request merged**